File: app/src/main/java/com/hwonchul/movie/domain/usecase/details/GetMovieUseCaseImpl.java
package com.hwonchul.movie.domain.usecase.details;

import androidx.annotation.NonNull;

import com.hwonchul.movie.domain.model.MovieDetail;
import com.hwonchul.movie.domain.repository.MovieRepository;

import javax.inject.Inject;

import io.reactivex.rxjava3.core.Flowable;

public class GetMovieUseCaseImpl implements GetMovieUseCase {
    private final MovieRepository repository;

    @Inject
    public GetMovieUseCaseImpl(@NonNull MovieRepository repository) {
        this.repository = repository;
    }

    @Override
    public Flowable<MovieDetail> invoke(int movieId) {
        return repository.getMovieDetailById(movieId);
    }
}

File: app/src/main/java/com/hwonchul/movie/domain/usecase/details/GetMovieUseCaseImpl.kt
package com.hwonchul.movie.domain.usecase.details

import com.hwonchul.movie.domain.model.MovieDetail
import com.hwonchul.movie.domain.repository.MovieRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.map
import javax.inject.Inject

class GetMovieUseCaseImpl @Inject constructor(
    private val repository: MovieRepository
) : GetMovieUseCase {

    override suspend fun invoke(movieId: Int): Flow<Result<MovieDetail>> {
        return repository.getMovieDetailById(movieId)
            .map { Result.success(it) }
            .catch { emit(Result.failure(it)) }
    }
}